### PR TITLE
[Bugfix] [SQL-Search] Fix SQL code generation when :not operator is included.

### PR DESCRIPTION
**Pull request by fhibf**
## Description
Queries with the :not operator were returning more results than the expected.
When reviewing the existing logic, it was noticed that queries with the :not operator were not taking in consideration previous CTE than the one with the :not clause.

After changing the SearchParamTableExpressionKind to Normal (instead of All), the CTE usage was fixed. 

## Related issues
Addresses [issue #].

## Testing
- Unit Tests validation.
- E2E Tests validation. 
- Ad-hoc test validation. 

## FHIR Team Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [x] Tag the PR with Azure API for FHIR if this will release to the Azure API for FHIR managed service (CosmosDB or common code related to service)
- [x] Tag the PR with Azure Healthcare APIs if this will release to the Azure Healthcare APIs managed service (Sql server or common code related to service)
- [x] CI is green before merge
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: global.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
     "sdk": {
-        "version": "6.0.302"
+        "version": "6.0.400"
```
Comment (fhibf) [inline global.json:3]: Can I add this dotnet version upgrade to this PR?

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/NotExpressionTests.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+                string query3 = $"_id={practitioner.Id}&active:not=true";
+                Bundle queryResult3 = await _client.SearchAsync(ResourceType.PractitionerRole, query3);
+                Assert.Empty(queryResult3.Entry);
+            }
+
+            var healthCareServices = healthRecordIdentifiers.Where(i => i.ResourceType == ResourceType.HealthcareService.ToString());
+            foreach (HealthRecordIdentifier healthCareService in healthCareServices)
+            {
+                // Investigating the possibility of having a bug with :not and chain queries.
```
Comment (fhibf) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/NotExpressionTests.cs:61]: Additional investigation that I'm running: is it possible to have a chain query expression with the :not operator?

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/NotExpressionTests.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        public async Task GivenANotExpressionPattern_WhenSearched_ThenCompareResponseWithExpectedResults()
+        {
+            IReadOnlyList<HealthRecordIdentifier> healthRecordIdentifiers = await GetHealthRecordIdentifiersAsync(CancellationToken.None);
+
+            var practitionerRoles = healthRecordIdentifiers.Where(i => i.ResourceType == ResourceType.PractitionerRole.ToString());
+            foreach (HealthRecordIdentifier practitioner in practitionerRoles)
+            {
+                string query1 = $"_id={practitioner.Id}&active:not=false";
+                Bundle queryResult1 = await _client.SearchAsync(ResourceType.PractitionerRole, query1);
+                Assert.Single(queryResult1.Entry);
```
Comment (PTaladay) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/NotExpressionTests.cs:47]: Would it be a good idea to compare the expected single entry data wise to make sure it is the entry we expect not just that we only return one result?

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/NotExpressionTests.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+using Microsoft.Health.Test.Utilities;
+using Xunit;
+using ResourceType = Hl7.Fhir.Model.ResourceType;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.Health.Fhir.Tests.E2E.Rest.Search
+{
+    [Trait(Traits.Category, Categories.Search)]
+    [HttpIntegrationFixtureArgumentSets(DataStore.SqlServer, Format.Json)]
+    public sealed class NotExpressionTests : IClassFixture<HttpIntegrationTestFixture>
```
Comment (LTA-Thinking) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/NotExpressionTests.cs:27]: Should missing be added to the class/test names? It looks like that expression is also being tested in this file.

Comment (fhibf) [reply test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/NotExpressionTests.cs:27]: That's a good question. 

Originally the bug that I'm fixing is ":not operator not working as expected when combined with other operators.".
The addition of the "missing" operator is just to have the additional operator to force the problem.

The target of the test is not testing the :missing operator, but the usage of :not operator.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/NotExpressionTests.cs [hunk 5]
```diff
@@ -58,15 +60,15 @@ public async Task GivenANotExpressionPattern_WhenSearched_ThenCompareResponseWit
             var healthCareServices = healthRecordIdentifiers.Where(i => i.ResourceType == ResourceType.HealthcareService.ToString());
             foreach (HealthRecordIdentifier healthCareService in healthCareServices)
```
Comment (LTA-Thinking) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/NotExpressionTests.cs:61]: How is this part of the test relevant to the :not operator? Neither of the queries use it.

Comment (fhibf) [reply test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/NotExpressionTests.cs]: Moved those tests to a new class called IdSearchTests. 
That's another side-effect bug that was somehow related to a failures in :not operator.

**Review by LTA-Thinking — APPROVED** [review]

**Pull request merged**